pager/src/main/java/dev/chrisbanes/accompanist/pager/PagerState.kt

```diff
@@ -0,0 +1,396 @@
+            targetValue = page + pageOffset,
+            initialVelocity = initialVelocity,
+            animationSpec = animationSpec
+        ) { value, _ ->
+            currentPage = floor(value).toInt()
+            currentPageOffset = value - currentPage
+        }
+        snapToNearestPage()
+    }
+
+    private fun determineSpringBackOffset(
+        velocity: Float,
+        offset: Float = currentPageOffset,
+    ): Float = when {
+        // If the offset exceeds the scroll threshold (in either direction), we want to
+        // move to the next/previous item
+        offset < ScrollThreshold -> 0f
+        offset > 1 - ScrollThreshold -> 1f
+        // Otherwise we look at the velocity for scroll direction
+        velocity < 0 -> 1f
+        else -> 0f
+    }
+
+    internal val draggableState = DraggableState { delta ->
+        dragByOffset(delta / pageSize.coerceAtLeast(1))
+    }
+
+    private fun dragByOffset(deltaOffset: Float) {
+        val targetedOffset = currentPageOffset - deltaOffset
+
+        if (targetedOffset < 0) {
+            // If the target offset is < 0, we're trying to cross the boundary to the previous page
+            if (currentPage > 0) {
+                // We can only move to the previous page if we're not at page 0
+                currentPage--
+                currentPageOffset = targetedOffset + 1
+            } else {
+                // If we're at page 0, pin to 0f offset
+                currentPageOffset = 0f
+            }
+        } else if (targetedOffset >= 1) {
+            // If the target offset is > 1, we're trying to cross the boundary to the next page
+            if (currentPage < pageCount - 1) {
+                // We can only move to the next page if we're not on the last page
+                currentPage++
+                currentPageOffset = targetedOffset - 1
+            } else {
+                // If we're on the last page, pin to 0f offset
+                currentPageOffset = 0f
+            }
+        } else {
+            // Otherwise, we can use the offset as-is
+            currentPageOffset = targetedOffset
+        }
+
+        if (DebugLog) {
+            Log.d(
+                LogTag,
+                "dragByOffset. delta:%.4f, targetOffset:%.4f, new-page:%d, new-offset:%.4f"
+                    .format(deltaOffset, targetedOffset, currentPage, currentPageOffset),
+            )
+        }
+    }
+
+    /**
+     * TODO make this public?
+     */
+    internal suspend fun performFling(
+        initialVelocity: Float,
+        animationSpec: DecayAnimationSpec<Float>,
+    ) = draggableState.drag {
+        selectionState = SelectionState.Settling
+
+        // We calculate the target offset using pixels, rather than using the offset
+        val targetOffset = animationSpec.calculateTargetValue(
+            initialValue = currentPageOffset * pageSize,
+            initialVelocity = initialVelocity * -1
```

Comment:
Will this work for rtl? I assume it's a responsibility of draggable/scrollable to flip velocity for you if necessary

Reply:
Yeah, I'm not entirely sure why this is necessary (I didn't need this when doing everything manually). I'm going to take a pass through the fling code once we move to `scrollable()`.